src/bigistore.c

```diff
@@ -1099,6 +1099,65 @@ Datum bigistore_slice_to_array(PG_FUNCTION_ARGS)
     PG_RETURN_POINTER(aout);
 }
 
+static BigIStore * bigistore_clamp_pass(BigIStore * is, int32 clamp_key, int delta_dir)
+{
+    BigIStore     * result_is;
+    BigIStorePair * pairs;
+    int32 clamp_sum = 0;
+    int   index     = 0, count = 0, delta_buflen = 0, new_size = 0, bytesize = 0;
+
+    /* short circuit out of the funciton if there is nothing to clamp */
+    if(delta_dir > 0 && FIRST_PAIR(is, BigIStorePair)->key >= clamp_key)
+        return is;
+    if(delta_dir < 0 && LAST_PAIR(is, BigIStorePair)->key <= clamp_key) 
+        return is;
+
+    pairs = FIRST_PAIR(is, BigIStorePair);
+    index = delta_dir > 0 ? 0 : is->len - 1;
+    while ( ((delta_dir > 0) && (index < is->len && pairs[index].key <= clamp_key)) ||
+            ((delta_dir < 0) && (index >= 0      && pairs[index].key >= clamp_key)) )
+    {
+        INTPL(pairs[index].val, clamp_sum, clamp_sum);
+        delta_buflen += bigis_pair_buf_len(pairs + index);
+        index += delta_dir, count++;
+    }
+
+    /* back to the last element that is to be clamped */
+    index -= delta_dir, count--;
+
+    /* copy survivors into a new spot */
+    if (delta_dir > 0)
+        pairs = pairs + index;
+
+    new_size       = is->len - count;
+    bytesize       = ISHDRSZ + new_size * sizeof(BigIStorePair);
+    result_is      = palloc0(bytesize);
+    result_is->len = new_size;
+
```

Comment:
that looks like `FINALIZE_ISTORE` / `FINALIZE_BIGISTORE`

Reply:
Only this doesn't depend on having intermediate pairs structure and array allocated.

Reply:
Can wrap the existing pairs into that struct though...

Reply:
`FINALIZE_ISTORE` tries to `pfree` `pairs` in the end.

Reply:
I introduced an intermediate `FINALIZE_ISTORE_BASE` in the next commit that doesn't `pfree`

Reply:
what's wrong with the `pfree` ?

Reply:
nvm a saw it already